libcudacxx/include/cuda/__barrier/barrier_block_scope.h

```diff
@@ -381,12 +382,30 @@ class barrier<thread_scope_block, ::cuda::std::__empty_completion> : public __bl
 public:
   _CCCL_API inline void wait(arrival_token&& __phase) const
   {
+    // no need to back off on a barrier in SMEM on SM90+, SYNCS unit is taking care of this
+    NV_IF_TARGET(NV_PROVIDES_SM_90,
+                 (if (::cuda::device::is_object_from(__barrier, ::cuda::device::address_space::shared)) {
+                   while (!::cuda::ptx::mbarrier_try_wait(
+                     reinterpret_cast<uint64_t*>(const_cast<__barrier_base*>(&__barrier)), __phase))
+                     ;
+                   return;
+                 }))
```

Comment:
So it turns out the compiler could not optimize:
```c++
while (!__try_wait(__phase))
  ;
```
so I had to drop down to cuda::ptx.